### DISCUSSION TITLE
Why can't WalletClient Transport be passed into PublicClient?

### DISCUSSION THREAD

**Discussion by kasparkallas** [wevm/viem#731] [Question]
If I want to use the same transport as the wallet, i.e. wait for transactions through the wallet RPC, why is this syntax not supported?

```ts
const walletClient = createWalletClient({
  chain: mainnet,
  transport: custom(window.ethereum)
})

const publiClient = createPublicClient({ 
  chain: walletClient.chain,
  transport: walletClient.transport
})
```

**Comment by kasparkallas** [comment]
I'm wondering about this topic for 2 reasons:
* I want my dApp to be in sync with the wallet when waiting for transactions.
* Gnosis Safe has custom RPC call handling in their [SafeAppProvider](https://github.com/safe-global/safe-apps-sdk/blob/0cf10f735e0ef434613d431411fd581e9a770467/packages/safe-apps-provider/src/provider.ts#L7) which the wagmi [SafeConnector](https://github.com/wagmi-dev/references/blob/746df4f7f0b19dd8ca59e0cc7b6e53aa296bc767/packages/connectors/src/safe.ts) uses. In the SafeAppProvider, notice `eth_getTransactionByHash` & `eth_getTransactionReceipt` calls going through their SDK.

**Comment by jxom** [comment]
Hmm yeah, agree folks might want to reuse a transport from another Client for specific reasons, and I don’t think we should stop them from doing this.

Makes sense for your proposal to work though. Shouldn’t take too long to add in – hold tight!

**Reply by kasparkallas** [reply]
Thanks for the quick response! I added some additional context just now in another comment while you were possibly posting.

**Reply by jxom** [reply]
We were thinking of adding a `decorators` pattern for Clients, which will be useful for the cases where you may want to attach Actions from other Clients, ie:

```ts
const client = createWalletClient({
  chain: mainnet,
  decorators: [publicActions],
  transport: custom(window.ethereum)
})
```

Meaning you wouldn’t have to use two clients. However, this should be use sparingly as some Wallets may not support common public RPC methods, or may only support them once a user is connected. See: https://viem.sh/docs/faq.html#why-doesn-t-wallet-client-support-public-actions

**Comment by kasparkallas** [comment]
Actually I realized that it compiles like this:
```ts
const publicClient = createPublicClient({
  chain: walletClient.chain,
  transport: custom(walletClient.transport)
});
```

**Comment by kasparkallas** [comment]
As a bonus, using viem's `custom` and ethers' experimental [EIP1193Bridge](https://docs.ethers.org/v5/api/experimental/#experimental-eip1193bridge), it looks like I can also turn any ethers `Signer` or `Provider` into any viem client:
```ts
import { Eip1193Bridge } from "@ethersproject/experimental";

const provider = new ethers.providers.JsonRpcProvider();
const signer = provider.getSigner();

const eip1193Bridge = new Eip1193Bridge(signer, signer.provider);

const publicClient = createPublicClient({
  transport: custom(eip1193Bridge),
});

const walletClient = createWalletClient({
  transport: custom(eip1193Bridge),
});
```
